### PR TITLE
[MCDU] Add page identifier and Progress Page OPT and REC MAX functionality

### PR DESCRIPTION
**Pull request by MisterChocker**
## Summary of Changes

- Calculate REC MAX FL
- Auto-update PROG PAGE (also serves as demo for newly implemented page id system)
- Create PageID system for all mcdu pages
- OPT dummy with display logic
- Added delay presets

The page id system and delay presets help #1524 to create timed events by adding the ability to check for a specific mcdu page.
The example below shows an update interval of the PROG PAGE of 5 seconds, this is done to refresh the pages content to show dynamic changes on the page.

**Try a demo of whats possible**
I have set up a [demo branch](https://github.com/MisterChocker/a32nx/tree/delay_demo) to show some work on delays
_Added init a and init b delay to switch pages, added delay to inputs based on this [video](https://youtu.be/EoJ9uqMYH_k?t=17)_

### Important
The auto refresh can only exist once in every class and needs to be in the main scope.

### Auto refresh page
In order to prevent stacking, we write the ID to a variable which gets automatically checked and tries to delete the timeout when leaving or reloading the page early.
```js
mcdu.page.SelfPtr = setTimeout(() => {
    if (mcdu.page.Current === mcdu.page.ProgressPage) {
        CDUProgressPage.ShowPage(mcdu);
    }
}, mcdu.PageTimeout.Prog);
```

### Button event
If you want to simulate delay on e.g. some calculation on an entry, do the following.

Before:
```js
mcdu.onLeftInput[5] = () => {
    mcdu._cruiseEntered = true;
    const value = mcdu.inOut;
    mcdu.clearUserInput();
    if (mcdu.setCruiseFlightLevelAndTemperature(value)) {
        CDUInitPage.ShowPage1(mcdu);
    }
};
```

After:
Clear the scratchpad first, so the user can continue with other work.
It is important that the original condition maintains and the page check has control over whether to change page or not.

```js
mcdu.onLeftInput[5] = () => {
    mcdu._cruiseEntered = true;
    const value = mcdu.inOut;
    mcdu.clearUserInput();
    setTimeout(() => {
        if (mcdu.setCruiseFlightLevelAndTemperature(value)) {
            if (mcdu.page.Current === mcdu.page.InitPageA) {
                CDUInitPage.ShowPage1(mcdu);
            }
        }
    }, mcdu.getDelaySwitchPage());
};
```

On every page change the `clearDisplay()` method gets called and sets the `Current` page id to `Clear` aka `0`.
The `Current` gets updated on every page below `mcdu.clearDisplay();` with `mcdu.page.Current = mcdu.page.Name;`.
Replace `Name` with the corresponding name from the table below.
If creating new Pages, simply at to the bottom with last ID+1.

| File | Name | ID | Description |
| --- | --- | --- | --- |
| _none_ | `Current` | ID of currently shown page | Will always hold the currently open page ID, if page is set up |
| _none_ | `Clear` | `0` | Will automatically reset the page ID on page change |
| A320_Neo_CDU_AirportsMonitor | `AirportsMonitor` | `1` | |
| A320_Neo_CDU_AirwaysFromWaypointPage | `AirwaysFromWaypointPage` | `2` | |
| A320_Neo_CDU_AirwaysFromWaypointPage | `AirwaysFromWaypointPageGetAllRows` | `3` | |
| A320_Neo_CDU_AvailableArrivalsPage | `AvailableArrivalsPage` | `4` | |
| A320_Neo_CDU_AvailableArrivalsPage | `AvailableArrivalsPageVias` | `5` | |
| A320_Neo_CDU_AvailableDeparturesPage | `AvailableDeparturesPage` | `6` | |
| A320_Neo_CDU_AvailableFlightPlanPage | `AvailableFlightPlanPage` | `7` | |
| A320_Neo_CDU_DataIndexPage | `DataIndexPage1` | `8` | |
| A320_Neo_CDU_DataIndexPage | `DataIndexPage2` | `9` | |
| A320_Neo_CDU_DirectToPage  | `DirectToPage` | `10` | |
| A320_Neo_CDU_FlightPlanPage | `FlightPlanPage` | `11` | |
| A320_Neo_CDU_FuelPredPage | `FuelPredPage` | `12` | |
| A320_Neo_CDU_GPSMonitor | `GPSMonitor` | `13` | |
| A320_Neo_CDU_HoldAtPage | `HoldAtPage` | `14` | |
| A320_Neo_CDU_IdentPage | `IdentPage` | `15` | |
| A320_Neo_CDU_InitPage | `InitPageA` | `16` | |
| A320_Neo_CDU_InitPage | `InitPageB` | `17` | |
| A320_Neo_CDU_IRSInit | `IRSInit` | `18` | |
| A320_Neo_CDU_IRSMonitor | `IRSMonitor` | `19` | |
| A320_Neo_CDU_IRSStatus | `IRSStatus` | `20` | |
| A320_Neo_CDU_IRSStatusFrozen | `IRSStatusFrozen` | `21` | |
| A320_Neo_CDU_LateralRevisionPage | `LateralRevisionPage` | `22` | |
| A320_Neo_CDU_MenuPage | `MenuPage` | `23` | |
| A320_Neo_CDU_NavaidPage | `NavaidPage` | `24` | |
| A320_Neo_CDU_NavRadioPage | `NavRadioPage` | `25` | |
| A320_Neo_CDU_NewWaypoint | `NewWaypoint` | `26` | |
| A320_Neo_CDU_PerformancePage | `PerformancePageTakeoff` | `27` | |
| A320_Neo_CDU_PerformancePage | `PerformancePageClb` | `28` | |
| A320_Neo_CDU_PerformancePage | `PerformancePageCrz` | `29` | |
| A320_Neo_CDU_PerformancePage | `PerformancePageDes` | `30` | |
| A320_Neo_CDU_PerformancePage | `PerformancePageAppr` | `31` | |
| A320_Neo_CDU_PerformancePage | `PerformancePageGoAround` | `32` | |
| A320_Neo_CDU_PilotsWaypoint | `PilotsWaypoint` | `33` | |
| A320_Neo_CDU_PositionFrozen | `PosFrozen` | `34` | |
| A320_Neo_CDU_PositionMonitorPage | `PositionMonitorPage` | `35` | |
| A320_Neo_CDU_ProgressPage | `ProgressPage` | `36` | |
| A320_Neo_CDU_ProgressPage | `ProgressPageReport` | `37` | |
| A320_Neo_CDU_ProgressPage | `ProgressPagePredictiveGPS` | `38` | |
| A320_Neo_CDU_SelectedNavaids | `SelectedNavaids` | `39` | |
| A320_Neo_CDU_SelectWptPage | `SelectWptPage` | `40` | |
| A320_Neo_CDU_VerticalRevisionPage | `VerticalRevisionPage` | `41` | |
| A320_Neo_CDU_WaypointPage | `WaypointPage` | `42` | |

Page Timeouts
| method | update interval ms | Description |
| --- | --- | --- |
| `mcdu.PageTimeout.Prog` | 5000 | Defines the delay before page gets reloaded |

Methods created in advance
| method | ~delay ms | Description |
| --- | --- | --- |
| `getDelaySwitchPage()` | 150-200 | Used for switching pages
| `getDelayBasic()` | 300-400 | Used for basic inputs e.g. alternate airport, ci, fl, temp, constraints, ...
| `getDelayMedium()` | 600-800 | Used for e.g. loading time fore pages
| `getDelayHigh()` | 900-1200 | Used for intense calculation
| `getDelayRouteChange()` | 300-2000+ _(dynamic)_ | Used for changes to the flight plan
| `getDelayFuelPred()` | waypoint count ^ 2 + route lenght | Used for calculation time for fuel pred page
| `getDelayWindLoad()` | waypoint count ^ 2 | Used to load wind data into fms

## Screenshots (if necessary)

image

## References

[Reference for delay presets](https://www.youtube.com/watch?v=EoJ9uqMYH_k&t=115s)
image
image

## Additional context

Discord username (if different from GitHub): DerL30N#3751

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

### CODE REVIEW

**Comment by MisterChocker** [comment]
@Benjozork I am already on latest master, or is there a version I cannot merge yet?

**Comment by Benjozork** [comment]
The label was added automatically, removing it as people confirm or rebase.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js [hunk 1]
```diff
@@ -316,6 +316,51 @@ class FMCMainDisplay extends BaseAirliners {
         this._inOutElement.style.color = color;
     }
 
+    /**
+     * Returns true if an engine is not running (FF === 0)
+     * @returns {number}
+     */
+    isAnEngineOff() {
+        return Simplane.getEngineActive(0) && Simplane.getEngineActive(1);
```
Comment (Lucky38i) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js:324]: shouldn't this be an or? looks like it checks if both engines are active

Comment (MisterChocker) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js:324]: Yes to match the name it should be, the method should be called `areBothEnginesOn`

Comment (MisterChocker) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js:324]: Fixed

Comment (Lucky38i) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js:324]: doesn't the FCOM state that an engine needs to be one, not both?

Comment (MisterChocker) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js:324]: this is used to check whether an engine fail or happened, this is a condition whether to shop OPT FL or not.

**Comment by Benjozork** [comment]
Hey, can you provide a quick list of things QA can test ?

**Comment by MisterChocker** [comment]
@Benjozork yes sure.

- Use the weather menu and change the temperature, monitor the prog page, update interval should be around 5 seconds.
Check the REC MAX alt for deviation with the chart above.
- Check if you can manage to stack events on the prog page. (this should not be possible)
If the update interval of the prog page gets faster, which it shouldn't, you managed to stack events.
You can do so by switching from and to the prog page and try to add events, meanwhile you can change the temperature in the weather menu and check for changes on REC MAX.

Basically change back and fort to prog page while changing the temp to see rate of change by changes of REC MAX.

**Comment by danbw9** [comment]
Doing a flight with this now. Nice addition! @MisterChocker I wanted to confirm that your 'dummy' for OPT should still show a FL? At the movement I have white dashes. REC Max varying correctly with temperature and weight, cross checked via the graph references above. 

Screenshot (625)

**Comment by MisterChocker** [comment]
@danbw9 yes you are right, my' dummy' OPT should display the same FL as the REC MAX.
Did you set block fuel and cruise fl? Also do you have both engines running?

**Comment by z0rgmeister** [comment]
Name: z0rgmeister#8972
Date of testing: 07/11/2020
Version of the sim: 1.10.8.0
PR Tested: #1772
Object of test: To test Progress Page OPT and REC MAX
Steps taken: 

1.Full LPFR-LPPT flight from C&D.
2.Verified through use the PROG Page functionality as per PR.
3.Verified dummy OPT display in PROG Page not displaying any FL.


Flight notes: NA

Taxi/Takeoff : NA

Climb/Cruise: NA

Descent/Approach/Landing: DNA

Shutdown: NA

Issues: Not sure it's an issue. OPT not showing any FL just dashes.

Overall Rating: Good

Conclusions of the testing: Good to go, nice addition. Not sure the dashes in OPT are intended or not. Not major.

**Comment by danbw9** [comment]
> @danbw9 yes you are right, my' dummy' OPT should display the same FL as the REC MAX.
> Did you set block fuel and cruise fl? Also do you have both engines running/operational?
> 
> EDIT: Found an error in my condition, corrected it push now.

Thanks for correcting this. Will try your new commit on my next flight and report back.

**Comment by Carter633** [comment]
Ran a test and had a similar result to Dan - without pumping a huge QA template here I'll review again tomorrow afternoon (after 1400Z) if it's still open to test with the amendments

It's yet another amazing addition though!

BritishAvgeek YT - Carter633#1436

**Comment by MisterChocker** [comment]
Thanks guys! I sneaked one more change in ^^.
Now the OPT FL shows the REC MAX FL but in 500 ft increments. I have set it to floor. If REC MAX is 359, OPT will show 355.

**Comment by MisterChocker** [comment]
looking good ;) 
image
image
One engine out
image

**Comment by danbw9** [comment]
Name: danbw#7028
Date of testing: 07/11/2020
Version of the sim: 1.10.8.0
PR Tested: #1772
Object of test: Check out PROG Page OPT and REC MAX
Steps taken:

Full flight from C&D - UACC to UAAA. 

Observing the REC MAX value on the ground and noting the environmental conditions at those stages of flight and compare to the graph in references above. Largely all correct. 

Noted that OPT CLB and REC MAX which are always the same FL (for the time being) increase with GW weight decreasing (assuming same temperature)

The 'REPORT' function on PROG page now works, which when last time I tried did not work. Not sure if this was 'enabled' along with the PROG page functionality.

Flight notes: Noting temperature along with GW what gives values in PROG page - considering ISA deviation too. 

Taxi/Takeoff : NA

Climb/Cruise: REC MAX value increasing throughout CRZ as appropriate. Does not exceed FL390 even in a super light configuration.

Descent/Approach/Landing: N/A

Shutdown: NA

Issues: None, current functionality is as described in PR
Overall Rating: Good
Conclusions of the testing: LGTM. The previous issue I reported regarding the white dashes on the under OPT CLB are gone. No issues. PR works all fine. Thank you for your work. As far as I remember from in IRL, the OPT CLB FL should already be available in the pre flight phase on the ground whilst setting up. This is something I would usually brief. At the moment on the ground it shows dashes.

**Comment by MisterChocker** [comment]
> As far as I remember from in IRL, the OPT CLB FL should already be available in the pre flight phase on the ground whilst setting up.

Corrected. OPT FL now visible when on ground as well after block fuel and zfw/cg are set.

**Comment by Benjozork** [comment]
Minor changes, but no need for new RTT
